lib/polyfill/patchedmediakeys_20140218.js

```diff
@@ -0,0 +1,832 @@
+  this.eventManager_.listen(media, 'msneedkey',
+      /** @type {shaka.util.EventManager.ListenerType} */
+      (v20140218.onMsNeedKey_));
+
+  var self = this;
+  function setMediaKeysDeferred() {
+    media.msSetMediaKeys(self.nativeMediaKeys_);
+    media.removeEventListener('loadedmetadata', setMediaKeysDeferred);
+  }
+
+  // Wrap native HTMLMediaElement.msSetMediaKeys with Promise
+  try {
+    // IE11/Edge requires that readyState >=1 before mediaKeys can be set, so
+    // check this and wait for loadedmetadata if we are not in the correct state
+    if (media.readyState >= 1) {
+      media.msSetMediaKeys(this.nativeMediaKeys_);
+    } else {
+      media.addEventListener('loadedmetadata', setMediaKeysDeferred);
+    }
+
+    return Promise.resolve();
+
+  } catch (exception) {
+    return Promise.reject(exception);
+  }
+};
+
+
+
+/**
+ * An implementation of MediaKeySession.
+ *
+ * @constructor
+ * @param {MSMediaKeys} nativeMediaKeys
+ * @param {string} sessionType
+ * @implements {MediaKeySession}
+ * @extends {shaka.util.FakeEventTarget}
+ */
+shaka.polyfill.PatchedMediaKeys.v20140218.
+    MediaKeySession = function(nativeMediaKeys, sessionType) {
+  shaka.log.debug('v20140218.MediaKeySession');
+  shaka.util.FakeEventTarget.call(this, null);
+
+  // Native MediaKeySession, which will be created in generateRequest
+  /** @private @type {MSMediaKeySession} */
+  this.nativeMediaKeySession_ = null;
+
+  /** @private @type {MSMediaKeys} */
+  this.nativeMediaKeys_ = nativeMediaKeys;
+
+  // Promises that are resolved later
+  /** @private @type {Promise} */
+  this.generateRequestPromise_ = null;
+
+  /** @private @type {Promise} */
+  this.updatePromise_ = null;
+
+  /** @private {string} */
+  this.type_ = sessionType;
+
+  /** @type {string} */
+  this.sessionId = '';
+
+  /** @type {number} */
+  this.expiration = NaN;
+
+  /** @type {!shaka.util.PublicPromise} */
+  this.closed = new shaka.util.PublicPromise();
+
+  /** @type {!MediaKeyStatusMap} */
+  this.keyStatuses =
+      new shaka.polyfill.PatchedMediaKeys.v20140218.MediaKeyStatusMap();
+};
+goog.inherits(shaka.polyfill.PatchedMediaKeys.v20140218.MediaKeySession,
+    shaka.util.FakeEventTarget);
+
+
+/** @override */
+shaka.polyfill.PatchedMediaKeys.v20140218.MediaKeySession.prototype.
+    generateRequest = function(initDataType, initData) {
+  shaka.log.debug('v20140218.MediaKeySession.generateRequest');
+
+  this.generateRequestPromise_ = new shaka.util.PublicPromise();
+
+  try {
+    // This EME spec version requires a MIME content type as the 1st param
+    // to createSession, but doesn't seem to matter what the value is...
+    this.nativeMediaKeySession_ = this.nativeMediaKeys_
+        .createSession('video/mp4', initData);
+
+    // Alias
+    var v20140218 = shaka.polyfill.PatchedMediaKeys.v20140218;
+
+    // Attach session event handlers here
+    this.nativeMediaKeySession_.addEventListener('mskeymessage',
```

Comment:
These event listeners are never cleaned up.  Use EventManager here and call removeAll() when the session is closed.